pkg/pillar/types/volumetypes.go

```diff
@@ -180,3 +180,47 @@ func (status VolumeStatus) LogDelete() {
 func (status VolumeStatus) LogKey() string {
 	return string(base.VolumeStatusLogType) + "-" + status.Key()
 }
+
+// VolumeRefConfig: Reference to a Volume specified separately in the API
+// If a volume is purged (re-created from scratch) it will either have a new
+// UUID or a new generationCount
+type VolumeRefConfig struct {
+	VolumeID          uuid.UUID
+	GenerationCounter int64
+	RefCount          uint
+}
+
+// Key: VolumeRefConfig unique key
+func (config VolumeRefConfig) Key() string {
+	return fmt.Sprintf("%v#%d#%d", config.VolumeID, config.GenerationCounter,
+		config.RefCount)
+}
+
+// VolumeKey: Unique key of volume referenced in VolumeRefConfig
+func (config VolumeRefConfig) VolumeKey() string {
+	return fmt.Sprintf("%v#%d", config.VolumeID, config.GenerationCounter)
+}
+
+// VolumeRefStatus: Reference to a Volume specified separately in the API
+// If a volume is purged (re-created from scratch) it will either have a new
+// UUID or a new generationCount
+type VolumeRefStatus struct {
+	VolumeID           uuid.UUID
+	GenerationCounter  int64
+	RefCount           uint
+	State              SwState
```

Comment:
Wouldn't you also want a progress integer to report? Or will the controller/UI know to get that from the volume?

Reply:
I think the controller will get progress from content tree and volume (based on the info messages definition)

Reply:
I would be good to check that with Ravi and Bharani that this is indeed the case.
If so we can remove some more code in a separate PR. Hmm might still need some of it for baseos download progress. So we should get the baseos to use volume API first; don't know if this is in the works for the cloud side.

Reply:
Yes, moving baseos to new volume API is already in my TODO list.

Reply:
Did you check that the controller/UI progress comes from the content tree/volume and not from the app instance?